### PR TITLE
Fix unbounded DTensor sharding propagation cache growth

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #178301

Fixes https://github.com/pytorch/pytorch/issues/178276

The C++ DTensor dispatch fast path (NativeShardingPropagatorCache)
hashes non-tensor items inside list arguments unconditionally into the
cache key. For foreach optimizer ops like `_foreach_div_.ScalarList` and
`_foreach_addcdiv_.ScalarList`, the ScalarList contains step-varying
values (AdamW bias corrections change every training step). This causes
a new cache entry on every step, leaking OpStrategy, OpSpec, and
OutputSharding objects indefinitely.

The fix: apply the same `static_argnum` filtering to non-tensor items
inside lists as is already applied to top-level non-tensor arguments.
Since foreach ops register with `static_argnum=100` (default), their
scalar list values are excluded from the cache key, and the cache
stabilizes after warmup.

Authored with Claude.

### CODE REVIEW

File: test/distributed/tensor/test_optimizers.py [hunk 1]
```diff
@@ -717,6 +719,72 @@ def _input_fn_2d(mod, inputs, device_mesh):
...
+
+        gc.collect()
+        gc.disable()
+        try:
+            before = Counter(type(o).__qualname__ for o in gc.get_objects())
+
+            for _ in range(10):
+                run_step()
+
+            gc.collect()
```
Comment (zpcore) [inline test/distributed/tensor/test_optimizers.py]: I think a simplified way to test is to use `hits, misses = _get_fast_path_sharding_prop_cache_stats()`. We should see `misses` keeps increasing.

**Review by zpcore — APPROVED** [review]
LGTM!

File: test/distributed/tensor/test_optimizers.py (hunk 1 above)
Comment (wconstab) [reply test/distributed/tensor/test_optimizers.py]: changed to this way

Comment (zpcore) [reply test/distributed/tensor/test_optimizers.py]: I didn't realize we have `_get_DTensor_sharding_propagator_cache_stats` :)

**Comment by huydhn** [comment]
@pytorchbot revert -m 'The distributed test failures look legit' -c ignoredsignal

test/distributed/tensor/test_optimizers.py::TestDTensorOptimizerWithLocalTensor::test_adamw_sharding_cache_no_leak [GH job link](https://github.com/pytorch/pytorch/actions/runs/23653386028/job/68907322529) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/3a42f8241b6a0a3a15ae0fc3fa1d0122c2f2b742)

**Review by zpcore — APPROVED** [review]
